libyul/AsmParser.cpp

```diff
@@ -131,7 +131,9 @@ void Parser::fetchSourceLocationFromComment()
 		return;
 
 	static regex const lineRE = std::regex(
-		R"~~~((^|\s*)@src\s+(-1|\d+):(-1|\d+):(-1|\d+)(\s+|$))~~~",
```

Comment:
Since we plan to parse more tags, would it be possible to use an instance of Scanner for this?

Reply:
Probably. I'd have to create a completely new scanner, just for the single comment that is being processed?

But I'd rather leave it as is here and do it in a new PR. It looks like it will completely change how this function works and I have this already tested and working so I'd rather not drag this PR out too much.

Reply:
Yes, but Scanner is a very simple class. We already do exactly the same thing for parsing `@use-src`,

Reply:
After trying to use the scanner, looks like ensuring that tags nested in code snippets are not treated as actual tags is problematic.

I could first use a regex to find all substrings containing `@src` tags with parameters and then process each with a scanner. But to be able to determine where each substring ends, the regex would look pretty much like the one I have now. It's not what you wanted, is it? A simpler regex would not skip nested `@src` in code snippets.

I could also run the scanner over the whole comment and only start processing a new tag once the current one ends. This looks like the right solution to me but the problem here is that the scanner does not have a token for `@` and I think that adding it would affect how the language is parsed (some `@` inside other stuff would now become standalone tokens?).

I see that the Natspec parser just manually finds `@` inside the string and only then processes the part between `@` and the end of a line as a natspec tag. I could do that if I could assume that there can be only one tag per line. Can we make such an assumption here? Currently multiple `@src` tags on a single line are accepted and the last one wins.

Reply:
Ok, this starts to get very complicated :)
I think we can make our own assumptions/restrictions as long as printing and parsing is in line.
Forcing the tags to be at the start of a line is probably a good idea. I hope there are no tools that do reformatting.
Somewhere I saw that the line breaks inside are not preserved - but maybe that was just the natspec parser.

Reply:
What about one tag per line assumption? Without either that or adding `@` token to the scanner it's still problematic.

> Somewhere I saw that the line breaks inside are not preserved - but maybe that was just the natspec parser.

The natspec parser stops at the end of the line as far as I can tell:
https://github.com/ethereum/solidity/blob/4615e625142b9c8366c87bc56b3c6f0fad2091d5/libsolidity/parsing/DocStringParser.cpp#L90-L95

Reply:
For reference, here's what we agreed on on the chat:
> can't we use a generic regex to search for tags first and then when we find the `@src` tag, we skip over the string?
> (and with tags I mean anything starting with `@`)

> no I mean just search for tags starting with `@`
> if you find one, but the tag is unknown, we ignore it and just search for the next `@`
> if there is a string, then it is wrong
> but it's also unknown tag
> so I think it's fine to properly consider the string for `@src`, and do nothing for the other tags (that don't also have strings as arguments)

Reply:
Here's a new PR that does this refactor: #11913.
The PR does not really have anything to do with snippets and I had a bunch of other refactors related to parsing `@src` tags to I extracted it all to that PR. For this reason I rebased the snippet PR on top of it so now it's a dependency.